lib/flutter_radio.dart

```diff
@@ -1,36 +1,176 @@
+    await FlutterRadio.setMeta(item);
+  }
 
-  static Future<void> audioStart() async {
-    await _channel.invokeMethod('audioStart');
+  static Future<void> playOrPause({@required String url}) async {
+    try {
+      if (FlutterRadio._isPlaying) {
+        FlutterRadio.pause(url: url);
+      } else {
+        FlutterRadio.play(url: url);
+      }
+    } catch (err) {
+      throw Exception(err);
+    }
   }
 
   static Future<void> play({@required String url}) async {
-    final Map<String, dynamic> params = <String, dynamic>{'url': url};
-    await _channel.invokeMethod('play', params);
+    try {
+      String result =
+          await _channel.invokeMethod('play', <String, dynamic>{
+        'path': url,
+      });
+      print('result: $result');
+
+      _setPlayerCallback();
+
+      if (FlutterRadio._isPlaying) {
+        throw Exception('Player is already playing.');
+      }
+      FlutterRadio._isPlaying = true;
+
+      return result;
+    } catch (err) {
+      throw Exception(err);
+    }
   }
 
   static Future<void> pause({@required String url}) async {
-    final Map<String, dynamic> params = <String, dynamic>{'url': url};
-    await _channel.invokeMethod('pause', params);
+    if (!FlutterRadio._isPlaying) {
+      throw Exception('Player already stopped.');
+    }
+    FlutterRadio._isPlaying = false;
+
+    String result = await _channel.invokeMethod('pause');
+    _removePlayerCallback();
+    return result;
   }
 
   static Future<void> stop() async {
-    await _channel.invokeMethod('stop');
+    if (!FlutterRadio._isPlaying) {
+      throw Exception('Player already stopped.');
+    }
+    FlutterRadio._isPlaying = false;
+
+    String result = await _channel.invokeMethod('stop');
+    _removePlayerCallback();
+    return result;
   }
 
   static Future<bool> isPlaying() async {
-    final bool playing = await _channel.invokeMethod('isPlaying');
-    return playing;
+    return Future.value(_isPlaying);
   }
 
-  static Future<void> playOrPause({@required String url}) async {
-    final Map<String, dynamic> params = <String, dynamic>{'url': url};
-    await _channel.invokeMethod('playOrPause', params);
+  static Future<void> _setPlayerCallback() async {
+    if (_playerController == null) {
+      _playerController = new StreamController.broadcast();
+    }
+
+    _channel.setMethodCallHandler((MethodCall call) {
+      switch (call.method) {
+        case "updateProgress":
+          Map<String, dynamic> result = jsonDecode(call.arguments);
+          _playerController.add(new PlayStatus.fromJSON(result));
+          break;
+        default:
+          throw new ArgumentError('Unknown method ${call.method}');
+      }
+    });
+  }
+
+  static Future<void> _removePlayerCallback() async {
+    if (_playerController != null) {
+      _playerController
+        ..add(null)
+        ..close();
+      _playerController = null;
+    }
+  }
+
+  static Future<String> setMeta(AudioPlayerItem item) async {
```

Comment:
In my case stream doesnt have any information of current media info. It is fetched from another service, so i need to update current meta externally. Probably should be implemented on Android version too.